### PR TITLE
[BPF] support for ipv4 options

### PR DESCRIPTION
**Pull request by tomastigera**
## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf-gpl/globals.h [hunk 1]
```diff
@@ -16,11 +16,16 @@ struct cali_tc_globals {
 	__be32 host_tunnel_ip;
 	__be32 flags;
 	__be16 wg_port;
+	__be16 __pad;
 	__u32 natin_idx;
 	__u32 natout_idx;
 	__u8 iface_name[16];
 	__u32 log_filter_jmp;
 	__u32 jumps[32];
+	/* Needs to be 32bit aligned as it is followed by scratch area for
```
Comment (fasaxc) [inline felix/bpf-gpl/globals.h:25]: should we make the scratch area explicit here `__u32 scratch[xxx]`  Just thinking that removes a special case for this map and ensures the alignment?

File: felix/bpf-gpl/skb.h [hunk 2]
```diff
@@ -70,7 +70,8 @@ static CALI_BPF_INLINE long skb_iphdr_offset(struct cali_tc_ctx *ctx)
 	if (CALI_F_IPIP_ENCAPPED) {
 		// Ingress on an IPIP tunnel: skb is [ether|outer IP|inner IP|payload]
 		// TODO: we need to consider different types of IPIP tunnels like 4in6 or 6in4
-		return sizeof(struct ethhdr) + ctx->ipheader_len;
+		// XXX no support for ip options in ipip heder
```
Comment (fasaxc) [inline felix/bpf-gpl/skb.h]:
```suggestion
		// XXX no support for ip options in ipip header
```

File: felix/bpf-gpl/tc.c [hunk 3]
```diff
@@ -1042,8 +740,30 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
 					ip_hdr(ctx)->protocol == IPPROTO_UDP ? BPF_F_MARK_MANGLED_0 : 0);
 		}
 
-		res |= bpf_l3_csum_replace(skb, l3_csum_off, state->ip_src, state->ct_result.nat_sip, 4);
-		res |= bpf_l3_csum_replace(skb, l3_csum_off, state->ip_dst, state->post_nat_ip_dst, 4);
+		if (!in_place) {
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:756]: Any value in wrapping this up as a utility function, pops up in a few places (but maybe calling a utility function would be almost as verbose...

**Review by fasaxc — COMMENTED** [review]
Feels a lot less scary after you talked me through the refactor.

I'm a little wary of copying the data out and into the packets again; seems like a bit of an overhead but perhaps that can be factored back out again.  Maybe we should just move wholesale to the bpf-helper-based load/store functions and abandon direct access.

File: felix/bpf-gpl/globals.h (hunk 1 above)
Comment (tomastigera) [reply felix/bpf-gpl/globals.h:25]: yup, made it explicit

File: felix/bpf-gpl/tc.c (hunk 3 above)
Comment (tomastigera) [reply felix/bpf-gpl/tc.c:756]: yeah, thought about it a long time ago, but not worth it.

**Pull request merged**